File: app/src/main/java/os/dtakac/feritraspored/ui/listener/DebouncedMenuItemClickListener.java
package os.dtakac.feritraspored.ui.listener;

import android.os.SystemClock;
import android.view.MenuItem;

public abstract class DebouncedMenuItemClickListener implements MenuItem.OnMenuItemClickListener {

    private long lastClickTime;
    private final long threshold;

    public DebouncedMenuItemClickListener(long threshold){
        this.threshold = threshold;
    }

    @Override
    public boolean onMenuItemClick(MenuItem menuItem) {
        if(SystemClock.elapsedRealtime() - lastClickTime < threshold){
            //consume click and prevent others from executing
            return true;
        }
        lastClickTime = SystemClock.elapsedRealtime();
        onDebouncedClick();
        return false;
    }

    public abstract void onDebouncedClick();
}

File: app/src/main/java/os/dtakac/feritraspored/common/view/debounce/DebouncedMenuItemClickListener.kt
package os.dtakac.feritraspored.common.view.debounce

import android.os.SystemClock
import android.view.MenuItem
import os.dtakac.feritraspored.common.constants.DEBOUNCE_INTERVAL_MS

class DebouncedMenuItemClickListener(
        private val threshold: Long,
        private val listener: () -> Unit
): MenuItem.OnMenuItemClickListener {
    private var lastClickTime: Long = 0

    override fun onMenuItemClick(menuItem: MenuItem?): Boolean {
        if(SystemClock.elapsedRealtime() - lastClickTime < threshold) {
            //consume click and prevent others from executing
            return true
        }
        lastClickTime = SystemClock.elapsedRealtime()
        listener.invoke()
        return false
    }
}

fun MenuItem.onDebouncedClick(debounceInterval: Long = DEBOUNCE_INTERVAL_MS, listener: () -> Unit) {
    setOnMenuItemClickListener(DebouncedMenuItemClickListener(debounceInterval, listener))
}